src/PureScript/Backend/Optimizer/Semantics.purs

```diff
@@ -1150,14 +1150,25 @@ caseNumber = case _ of
   NeutLit (LitNumber a) -> Just a
   _ -> Nothing
 
-type Ctx =
+newtype Ctx = Ctx
   { currentLevel :: Int
-  , lookupExtern :: Tuple (Qualified Ident) (Maybe BackendAccessor) -> Maybe (Tuple BackendAnalysis NeutralExpr)
+  , lookupExtern :: Qualified Ident -> Maybe String -> Maybe BackendAnalysis
+  , analyze :: Ctx -> BackendSyntax BackendExpr -> BackendAnalysis
   , effect :: Boolean
   }
 
 nextLevel :: Ctx -> Tuple Level Ctx
-nextLevel ctx = Tuple (Level ctx.currentLevel) $ ctx { currentLevel = ctx.currentLevel + 1 }
+nextLevel (Ctx ctx) = Tuple (Level ctx.currentLevel) $ Ctx ctx { currentLevel = ctx.currentLevel + 1 }
+
+effectfully :: Ctx -> Ctx
+effectfully (Ctx ctx)
+  | ctx.effect = Ctx ctx
+  | otherwise = Ctx ctx { effect = true }
+
+purely :: Ctx -> Ctx
+purely (Ctx ctx)
+  | ctx.effect = Ctx ctx { effect = false }
+  | otherwise = Ctx ctx
```

Comment:
Why bother checking the value of `effect`?

```suggestion
effectfully :: Ctx -> Ctx
effectfully (Ctx ctx) = Ctx ctx { effect = true }

purely :: Ctx -> Ctx
purely (Ctx ctx) = Ctx ctx { effect = false }
```

Reply:
To avoid allocating a new Ctx if it’s unnecessary.